### PR TITLE
Intrinsic exponentials

### PR DESCRIPTION
**Pull request by GiudGiud**
This PR introduces rational approximations to compute exponentials. These replace quadratic interpolation, are more accurate and faster. Thanks @cjosey for computing all these !
They are based on computing a G term with a fractional approximation, then rebuilding F1, F2 and H from G. This approach is the fastest, but not the most accurate.
- segment integration time goes down from 9.6 ns on the green cluster to 6.1 ns for 3D linear source
- absolute accuracy goes from roughly 5e-6 to 5e-7, over a large domain (tau in [0, 100]). 
- relative accuracy is more limited, see plots below. x (=tau_3D) is in [1e-6, 10]

F1 term :
Linear source exp_F1 term absolute error
Linear source exp_F1 term fractional error

F2 term :
Linear source exp_F2 term absolute error
Linear source exp_F2 term fractional error

H term :
Linear source exp_H term absolute error
Linear source exp_H term fractional error

Other exponential approximations are included if after more testing we decide the provided accuracy is not enough.
Almost all tests have had to be updated, but as you can see a lot of the fluxes are the same to 5-6 digits.

### CODE REVIEW

**Comment by GiudGiud** [comment]
I added the test suite fix (2 threads instead of 4 to avoid oversubscribing on Travis), and reduced the discretization in the two longer tests (test_forward_3D_lattice and test_forward_3D_lattice_linear) for convenience.

**Comment by GiudGiud** [comment]
@mit-crpg/openmoc-dev anyone to review this PR ?
I'll need it to generate results for the M&C paper (with the rotations' fix)

**Comment by wuwenbin2006** [comment]
I am glad to do it this weekend. Wenbin Wu On 03/14/2019 12:39, Giud wrote: @mit-crpg/openmoc-dev anyone to review this PR ? I'll need it to generate results for the M&C paper (with the rotations' fix) — You are receiving this because you are on a team that was mentioned. Reply to this email directly, view it on GitHub, or mute the thread.

File: src/exponentials.h [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+ *          However, accuracy when reconstructing F1, F2 and H is lower.
+ *          It was generated by Colin Josey using Remez's algorithm.
+ * @param x the value at which to evaluate the function, usually tau
+ * @param expH a pointer to the exponential
+ */
+inline void expG_fractional(FP_PRECISION x, FP_PRECISION* expG) {
+
+  /* Coefficients for numerator */
+  const FP_PRECISION p0 = 0.5;
+  const FP_PRECISION p1 = 1.76558112351595 * 1E-1;
```
Comment (wuwenbin2006) [inline src/exponentials.h]: Upper case or lower case E, just keep the same.

File: src/exponentials.h [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+ * @date January 16, 2019
+ * @author Guillaume Giudicelli, MIT, Course 22 (g_giud@mit.edu)
+ */
+
+#ifndef EXPONENTIALS_H_
+#define EXPONENTIALS_H_
+
+/**
+ * @brief Computes most of the F1 exponential term.
+ * @details This method computes (1-exp(-x))/x using a 7th order polynomial
```
Comment (wuwenbin2006) [inline src/exponentials.h]: Any excellent implemented tools for Remez's algorithm? I found one, but failed to compile it. https://github.com/samhocevar/lolremez

File: src/exponentials.h [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+  den = den*x + d3;
+  den = den*x + d2;
+  den = den*x + d1;
+  den = den*x + d0;
+  *expH = num/den;
+}
+
+
+/**
+ * @brief Evaluates the function 1 - exp(tau) for tau negative.  Vectorises well.
```
[the remark this answers is not in the document]
Comment (GiudGiud) [reply src/exponentials.h]: fixed

File: src/exponentials.h [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+ * @brief Math approximations for computing exponentials.
+ * @date January 16, 2019
+ * @author Guillaume Giudicelli, MIT, Course 22 (g_giud@mit.edu)
+ */
+
+#ifndef EXPONENTIALS_H_
+#define EXPONENTIALS_H_
+
+/**
+ * @brief Computes most of the F1 exponential term.
```
[the remark this answers is not in the document]
Comment (GiudGiud) [reply src/exponentials.h]: not much. deleted

File: src/exponentials.h (hunk 2 above)
Comment (GiudGiud) [reply src/exponentials.h]: @cjosey implemented his own as far as I know

File: src/exponentials.h (hunk 1 above)
Comment (GiudGiud) [reply src/exponentials.h]: Ok keeping the same convention for all routines now

File: src/CPULSSolver.cpp [hunk 5]
```diff
@@ -402,23 +386,46 @@ void CPULSSolver::tallyLSScalarFlux(segment* curr_segment, int azim_index,
...
+    FP_PRECISION exp_G[_num_groups] __attribute__ ((aligned(VEC_ALIGNMENT)));
+    FP_PRECISION tau[_num_groups] __attribute__ ((aligned(VEC_ALIGNMENT)));
+    FP_PRECISION inv_sin_theta = exp_evaluator->getInverseSinTheta();
+    FP_PRECISION length_2D = exp_evaluator->convertDistance3Dto2D(length);
+
+#pragma omp simd aligned(sigma_t, tau, exp_G)
+    for (int e=0; e < _num_groups; e++) {
+      /* Bound tau by 1e-6 to limit error on the F2 term */
+      tau[e] = std::max(FP_PRECISION(1e-6), length_2D * sigma_t[e]);
+      expG_fractional(inv_sin_theta * tau[e], &exp_G[e]);
```
[the remark this answers is not in the document]
Comment (GiudGiud) [reply src/CPULSSolver.cpp]: Well now that we dont call the expEvaluator we actually can do that yeah

File: src/exponentials.h [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+ * @brief Evaluates the function 1 - exp(tau) for tau negative.  Vectorises well.
+ *        Accurate to 6.18 digits (single precision) for entire domain (including
+ *        near zero, unlike intrinsic)
+ *
+ *        Valid for tau ~= [-1.5e6, 0]
+ * @param tau input
+ * @param expv output
+ * @param length length of vector
+ */
+inline void cram7(float x, FP_PRECISION* expv) {
```
[the remark this answers is not in the document]
Comment (GiudGiud) [reply src/exponentials.h]: fixed

**Comment by wuwenbin2006** [comment]
Please resolve the conflicts. Then it is good to merge.

**Comment by GiudGiud** [comment]
Done
Thanks a lot for reviewing. The additional simplification you mandated reduced integration time by .2 ns, to 5.9 ns !

File: src/CPULSSolver.h [hunk 7]
```diff
@@ -102,7 +109,9 @@ class CPULSSolver : public CPUSolver {
   void stabilizeFlux();
   void checkLimitXS(int iteration);
 
-  /* Linear source constant term computation routines */
+  FP_PRECISION getFluxByCoords(LocalCoords* coords, int group);
```
Comment (wuwenbin2006) [inline src/CPULSSolver.h]: Declared this twice in line 120

Comment (GiudGiud) [reply src/CPULSSolver.h]: fixed

File: src/exponentials.h (hunk 2 above)
Comment (cjosey) [reply src/exponentials.h]: Sorry for the slow reply.

I used Remez part of the way.  The precise order was:
1. Use [Boost Math](https://github.com/boostorg/math/blob/develop/minimax) minimax code (Remez) to get initial guess.
2. Use my own nonlinear optimization to tune the parameters

If you have access to Mathematica (I no longer do), the easiest I've found to work with is the [Function Approximation](https://reference.wolfram.com/language/FunctionApproximations/tutorial/FunctionApproximations.html) package.  It's significantly better than anything else I've used.

**Comment by wuwenbin2006** [comment]
I suppose updating the true_results is necessary to pass the test of linear_fixed_source.

**Pull request merged**